Text circuit drawer

Shobhit21287 opened the pull request:
# Summary
This PR addresses the feature request in #14713 and adds a draw function to rust which takes in ```&CircuitData``` type to give an ASCII art string representation for its visualisation. It currently focuses only on drawing ```StandardInstruction```, ```StandardGate``` and ```Unitary``` variants of the ```PackedOperation``` and is being added as a standalone feature rather than replacing the entire TextDrawer from the python space. 

```rust
struct VisualizationMatrix<'a> {
    layers: Vec<VisualizationLayer<'a>>,
    circuit: &'a CircuitData,
    clbit_map: Vec<usize>,
}

struct VisualizationLayer<'a>(Vec<VisualizationElement<'a>>);

enum VisualizationElement<'a> {
    #[default]
    Empty,
    VerticalLine(&'a PackedInstruction),
    Input(WireInputElement<'a>),
    DirectOnWire(OnWireElement<'a>),
    Boxed(BoxedElement<'a>),
}
```

```VisualizationMatrix```, ```VisualizationLayer``` struct and ```VisualizationElement``` enum are introduced to create a logical 2D matrix representation of the circuit instructions for visualization. 

1. ```fn build_layers(dag: &DAGCircuit) -> Vec<Vec<NodeIndex>>```:  This function is used to divide the ```NodeIndex``` of a ```DAGCircuit``` into layers so that when they are visualised, there are no instructions with overlapping wires.
2. ```fn get_instruction_range(...) -> (usize, usize)```: This function provides with range of indices of wires that will be affected by an instruction when it is visualised. It is used in ```build_layer``` to avoid overlapping of visualised instructions. 

Please note that these layers are purely for visualization. ```NodeIndex``` elements ```build_layers```out , that might be topologically the same, will most likely be in different layers, as their respective instructions act on intersecting ranges.

The ```VisualizationMatrix``` is an intermediate representation that handles all the logic mappings of instructions for visualization. This is then used to build another data struct, ```TextDrawer``` that handles all the string operations to create the final ASCII string art.

```rust
struct TextDrawer {
    wires: Vec<Vec<TextWireElement>>,
}

struct TextWireElement {
    top: String,
    mid: String,
    bot: String,
}
```
The ```TextDrawer``` has a ```draw_element``` method which gives the corresponding ```TextWireElement``` for a ```VisualizationElement```. The ```TextDrawer``` struct holds the 2D String representation which can then be printed using the ```.draw()``` method.

## Code flow

- The entry point takes in a ```CircuitData``` which is used to create the ```VisualizationMatrix```. Specifically for the ```build_layers``` function, ```CircuitData``` is converted to ```DAGCircuit``` 
- ```build_layers``` is used in ``` VisualizationMatrix::from_circuit()``` to construct the ```VisualizationMatrix``` for the input circuit
- Each element in the ```VisualizationMatrix``` is a logical ```VisualizationLayer```. This is a column based representation.
- Each layer holds the enum ```VisualizationElement```. These enums ultimately hold a reference to ```PackedInstruction```.
- After the ```VisualizationMatrix``` is created, it is then used to create ```TextDrawer```, which holds a row based representation for the circuit. These rows represent the wires for each qubit and clbit/register based on if ```cregbundle``` is true or not.
- Finally, post processing of the ASCII string art happens using the ```TextDrawer```, after which it returns the string. At this point it is ready for printing.

Shobhit21287 committed:
Entry point and function created, returns info about instructions in circuit

Shobhit21287 committed:
printing the qubits and clbits names

Shobhit21287 committed:
fixed entry point for draw function

Shobhit21287 committed:
addition to draw function, getting layers from DAG circuit

Shobhit21287 committed:
code moved to circuit_drawer.rs and added as module for import, no longer needed to import qiskit._accelerate, can simply call qc.draw(text2)

Shobhit21287 committed:
refactored code, wire and circuit class

Shobhit21287 committed:
minor qubit rep fixes

Shobhit21287 committed:
boilerplate code for layer stripping

Shobhit21287 committed:
creating visualisation layers

Shobhit21287 committed:
changed from pymodule to pyfunction

Shobhit21287 committed:
circuit to dag conversion moved to python

Shobhit21287 committed:
quadratic complexity building of sublayers

Shobhit21287 committed:
few minor fixes

Shobhit21287 committed:
get_layers implementation completed

Shobhit21287 committed:
datastruct for drawing rep

Shobhit21287 committed:
barrier implementation added

Shobhit21287 committed:
Visualization Matrix structure completed

eliarbel committed:
Make build_layers a standalone function

Shobhit21287 committed:
drawing logic for box and multibox, renamed Enclosed to Drawable

Shobhit21287 committed:
printing logic complete, normalisation left

Shobhit21287 committed:
fixed a counter in draw_layer, removed debug print statement to make code run

eliarbel committed:
Lay groundwork for VisualizationMatrix

Shobhit21287 committed:
general handling of standard gate and instructions added, inputs also handled

Shobhit21287 committed:
minor changes in print statements

Shobhit21287 committed:
Visualization Matrix implementation complete

Shobhit21287 committed:
Visualization Matrix implementation complete

Shobhit21287 committed:
text drawer, logic complete. TO DO: fix decorations

Shobhit21287 committed:
added swap gate support, removed circuit rep

Shobhit21287 committed:
added top and bottom connect support, fixed parameter labels

Shobhit21287 committed:
merge wires added

Shobhit21287 committed:
added cregbundle functionality

Shobhit21287 committed:
added draw capabilities

Shobhit21287 committed:
linting for lib.rs

Shobhit21287 committed:
added cregbundle flag

Shobhit21287 committed:
removed visualization matrix printing

Shobhit21287 committed:
mergewires flag added

eliarbel committed:
Clean up various `VisualizationMatrix` construction flow

eliarbel committed:
Run cargo fmt

Shobhit21287 committed:
fixed a few labels

Shobhit21287 committed:
added vertical lines support for controlled gates

Shobhit21287 committed:
added standard gate visualisation support barring special cases

Shobhit21287 committed:
fixed classical wire connection for measurement

Shobhit21287 committed:
added fold functionality

Shobhit21287 committed:
cosmetic addition for folding of quantum circuits

eliarbel committed:
Fix `cregbundle` behavior

The intention of the `cregbundle` parameter is to bundle the bits
of each classical register separately. This commit fixes a bug
where all classical register bits were bundled into a single classical
register.

Shobhit21287 committed:
added aspect ratio based folding for printing, index out of range bug for creg bundling cases

Shobhit21287 committed:
preliminary changes to structure, added PackedInstruction to control and swap, a function that tells the position of the direct on wire element and changed the code to use this function when deciding to use top and bottom connectors for boxed instructions.

Shobhit21287 committed:
bug fix for top and bottom con for boxed elements

eliarbel committed:
Simplify get_label

This commits refactors the `TextDrawer::get_label` method to be more
Rust idiomatic.

Shobhit21287 committed:
temporary commit to save changes before merging with text_circuit_drawer

eliarbel committed:
Implement Debug for VisualizationMatrix

eliarbel committed:
Simplify VisualizationElement::VerticalLine

eliarbel committed:
Add get_name to ElementWireInput

eliarbel committed:
Simplify DirectOnWire case

eliarbel committed:
Simplify the Boxed::Single(inst) case

eliarbel committed:
Simplify the Boxed::Multi case

eliarbel committed:
Fix clippy issues, refactor and clean up a bit

eliarbel committed:
Propagate "top" to `merge_wires`

eliarbel committed:
Apply miscellaneous fixes

* Change the VisualizationElement::VerticalLine to hold a PackedInstruction reference
* Fix capitalization of labels
* Fix clippy
* Correct drawing vertical line of a measure

eliarbel committed:
Fix minor visualization issues

This commit fixes several issues discovered via manual testing. Specifically:

*  Single bit wires - avoid printing `_1` for bits of length 1. For example `my_qr` instead of `my_qr_1` if `my_qr` is of length 1
* Changed some standard gate names to match those in the Python drawer
* Count number of chars in label to account for Unicode chars properly when computing label length.

eliarbel committed:
Fix bugs related to input wires handling

* Changed default names for qubit and clbit wires
* Added support for anonymous bits
* Fixed a bug with mapping clbit indices to `VisualizationMatrix` wires when `cregbunde=true`

eliarbel committed:
Add clbit_map as a field in VisualizationMatrix

Shobhit21287 committed:
added indexing for classical bits when bundling

Shobhit21287 committed:
cargo fmt

eliarbel committed:
Fix a panic when measuring anonymous clbits with cregbundle

eliarbel committed:
Add unitary gate support

Shobhit21287 committed:
linting and final checks

Shobhit21287 committed:
cleanups

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=15357) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=15357) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit?pullRequest=15357) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: eliarbel<br/>:x: Shobhit21287<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit?pullRequest=15357) it.</sub>

eliarbel committed:
Fix panic with mergewires=true and improve Delay drawing

Shobhit21287 committed:
Merge remote-tracking branch 'upstream/main' into text_circuit_drawer

Shobhit21287 committed:
fixed label extraction for packed instruction

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+
+                current_layer.as_mut().unwrap().push(node_index);
+            }
+        }
```
Just a small syntactical suggestion (since Jake always calls me out for writing overly indented code in my PRs). You can do something like this and save an indent:

```suggestion
            let NodeType::Operation(instruction_to_insert) = &dag.dag()[node_index]  else {
                continue;
            }
            let (node_min, node_max) = get_instruction_range(
                dag.get_qargs(instruction_to_insert.qubits),
                dag.get_cargs(instruction_to_insert.clbits),
                dag.num_qubits(),
            );
            // Check for instruction range overlap
            if (node_min..=node_max).any(|idx| used_wires.contains(&idx)) {
                current_layer = None; // Indication for starting a new layer
                used_wires.clear();
            }
            used_wires.extend(node_min..=node_max);

            if current_layer.is_none() {
                layers.push(Vec::new());
                current_layer = layers.last_mut();
            }

            current_layer.as_mut().unwrap().push(node_index);
        }
```

Shobhit21287 replied:
done in [5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+}
+
+/// Calculate the range (inclusive) of the given instruction qubits/clbits over the wire indices.
+/// The assumption is that clbits always appear after the qubits in the visualization, hence the clbit indices
```
Is this true if you use a custom wire order? Maybe the intent of this function will be more clear as I read through the PR but as a statement this assumption seems correct in the default case but not necessarily always true.

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+    let indices = node_qubits
+        .iter()
+        .map(|q| q.index())
+        .chain(node_clbits.iter().map(|c| c.index() + num_qubits));
```
Is there a reason you can't just do the map on the output instead of in an iterator for every element like this. In practice I would expect it to be compiled out so it doesn't matter too much, but it seems unnecessary.

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+            OperationRef::Unitary(_) => {
+                self.add_unitary_gate(inst, circuit);
+            }
+            _ => unimplemented!(
+                "{}",
+                format!(
+                    "Visualization is not implemented for instruction of type {:?}",
+                    inst.op
+                )
+            ),
```
Just a note for the future we have other rust native instruction types now, such as the pauli product measurement. It's not worth blocking this over that because we can add that incrementally very easily, but I wanted to make sure we don't forget. @Cryoris something to keep in mind for when you look at adding ppm to C

alexanderivrii replied:
Super-minor: I tried drawing a circuit with a `PauliProductRotation` before reading this comment or the PR summary, just to find out that its visualization is not yet supported. It might be a bit nicer to print `inst.op.name()` instead of `inst.op` though, as the message below is not very informative:

```
Visualization is not implemented for instruction of type PackedOperation(PackedOperationInner(0x121241cb4))
```

```suggestion
            _ => unimplemented!(
                "{}",
                format!(
                    "Visualization is not implemented for instruction of type {:?}",
                    inst.op.name()
                )
            ),
```

eliarbel replied:
Sasha: good suggestion. Addressed in 6f4f08d52e1ec016288f000e8d325f9077f93b35

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+        let (minima, maxima) = get_instruction_range(qargs, &[], 0);
+
+        match gate {
+            StandardGate::ISwap
+            | StandardGate::DCX
+            | StandardGate::ECR
+            | StandardGate::RXX
+            | StandardGate::RYY
+            | StandardGate::RZZ
+            | StandardGate::RZX
+            | StandardGate::XXMinusYY
+            | StandardGate::XXPlusYY
+            | StandardGate::RCCX
+            | StandardGate::RC3X => {
```
Are these contiguous ranges of u8 numbers in the enum definition. If they are it would be better to do this as an if, else if style thing so it would be something like:

```rust
if (12..=32).contains(gate as u8) {
     ...
} else if (0..12).contains(gate as u8) { 
    ...
}
```
since that will make the comparisons faster. Not that this is necessarily performance critical code.

If the ranges are not contiguous you can ignore the above comment (I didn't check them all).

Shobhit21287 replied:
The ranged are contiguous in terms of the number of qubits the gates need but not when it comes to visualization as a lot of 2 and 3 where a single qubit is being acted upon with the rest being control qubits, can be visualized in the same way but have `MultiBox` gates in between.

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+
+/// A Plain, logical 2D representation of a circuit.
+///
+/// A dense representation of the circuit of size N * (M + 1), where the first
```
`N` is the number of qubits and clibits/cregs right?

Shobhit21287 replied:
Yes, updated in [5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+            HashMap::with_capacity(circuit.data().len());
+        for (idx, node_index) in dag.op_node_indices(true).enumerate() {
+            node_index_to_inst.insert(node_index, &circuit.data()[idx]);
+        }
```
Why do you need this hashmap? Why can't you just get the instruction from the dag directly. That should be a faster access.

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+        let mut node_index_to_inst: HashMap<NodeIndex, &PackedInstruction> =
+            HashMap::with_capacity(circuit.data().len());
+        for (idx, node_index) in dag.op_node_indices(true).enumerate() {
+            node_index_to_inst.insert(node_index, &circuit.data()[idx]);
```
Is this correct? Shouldn't it be the dag you're accessing to get the packed instruction? While in practice doing the index mapping like this should work because the nodes are inserted into the dag in `circuit.data()` iteration order it seems kind of fragile to do this.

Shobhit21287 replied:
We wanted to use `CircuitData` as our primary reference for `PackedInstruction` as that output is close to what the user enters making testing easier for us and keeping the outputs fairly consistent. But we also needed the `DAGCircuit` to spool layer for drawing. That is why this approach was done. W.r.t the comment on why we need `HashMaps`, it's so that we can get rid of the `DAG` after mapping the `PackedInstruction` from it to those in the `CircuitData`, since there's no need for it after spooling.

If we don't care about the ordering in which the instructions are printed, we can probably get rid of `CircuitData` entirely and just use the `DAGCircuit`.

mtreinish replied:
I'm not sure I understand this reasoning. It shouldn't really matter as the topological ordering of the dag is fully deterministic so there should be no inconsistency in the drawing if just using the dag. But that doesn't really answer what I was asking here. I was more pointing out that assuming the node indices being in insertion order is an implementation detail and not a hard guarantee in the data structures. I think it would be more robust if we just used the dag.

mtreinish replied:
I removed the DAGCircuit usage in: https://github.com/Qiskit/qiskit/pull/15357/commits/ffc4ff2722dec8539b6c4c1e6ebb2304deb1231b this goes a bit further than what I was proposing initially but I think it's a cleaner implementation. The algorithm for layers construction is not exactly the same, but it's pretty close, the only big difference is it's sensitive to insertion order. We can improve or change this algorithm over time, but I think we should try to avoid the mixing of the two data structures.

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+                        .sum::<usize>()
+            };
+
+        let mut layers = vec![
+            VisualizationLayer(vec![VisualizationElement::default(); num_wires]);
+            inst_layers.len() + 1
+        ]; // Add 1 to account for the inputs layer
```
It functionally doesn't matter very much here especially since performance isn't critical but I wonder if there is a way to do this with an iterator collection instead so you avoid making all `VisualizationElement`s you will later just overwrite.

Shobhit21287 replied:
So we did this for convenience as this let us not care about the wires which don't have any operations. The elements that aren't being replaced default to empty wires. We can probably refactor it to collect all the `VisualizationElement`s but then we'd have to keep track of the indices that are empty. Is there be an efficient way to do this?

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+    }
+}
+
+// better name for the struct
```
Is this a todo?

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+impl TextWireElement {
+    fn width(&self) -> usize {
+        // return the max of all strings
+        // self.top.len().max(self.mid.len()).max(self.bot.len())
```
Is there a reason you didn't do this?

Shobhit21287 replied:
At the time this didn't work because ASCII characted take 3 bytes instead of 1 because `len()` returns the number of bytes in a string and not the characters. However, this can be easily fixed.

done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+
+    fn left_pad_string(s: &mut String, pad_char: char, width: usize) {
+        let current_width = s.len();
+        if current_width < width {
+            let pad_size = width - current_width;
+            let pad_str = pad_char.to_string().repeat(pad_size);
+            let new_str = format!("{}{}", pad_str, s);
+            *s = new_str;
+        }
+    }
+
+    fn pad_string(s: &mut String, pad_char: char, width: usize) {
+        let current_width = s.chars().count();
+        if current_width < width {
+            let pad_size = width - current_width;
+            let left_pad = pad_size / 2;
+            let right_pad = pad_size - left_pad;
+            let left_pad_str = pad_char.to_string().repeat(left_pad);
+            let right_pad_str = pad_char.to_string().repeat(right_pad);
+            let new_str = format!("{}{}{}", left_pad_str, s, right_pad_str);
+            *s = new_str;
+        }
+    }
```
I know I keep saying performance isn't critical for this functionality but I can't help myself. I feel like these padding functions are unnecessarily allocating a lot of intermediate strings. You should be able to do this by just growing the original string's allocation to account for the padding and not need so many additional allocations to add the padding characters.

Shobhit21287 replied:
updated in [5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+                match std_instruction {
+                    StandardInstruction::Measure => "M".to_string(),
+                    StandardInstruction::Reset => "|0>".to_string(),
+                    StandardInstruction::Barrier(_) => "░".to_string(),
```
You've got the constant defined did you want to use it here?
```suggestion
                    StandardInstruction::Barrier(_) => BARRIER.to_string(),
```

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+                    StandardGate::C3SX => "Sx",
+                    StandardGate::RC3X => "Rcccx",
+                }
+                .to_string();
```
I assume you're not doing this because the capitalization is different than the gate name (which are all lowercase iirc)?

```suggestion
                let mut label = if ![StandardGate::CSwap, StandardGate::Swap, StandardGate::GlobalPhaseGate].contains(standard_gate) {
                    standard_gate.name()
                } else {
                    ""
                }.to_string();
```

If that's the case and you need custom strings for the labels I'd be tempted to use a lookup table here like we do with names. So something like:
```
static STANDARD_GATE_LABELS: [&str; qiskit_circuit::operations::STANDARD_GATE_SIZE] = ["", "H", "I", "X", ...]
```
and then just make this:
```rust
let mut label = STANDARD_GATE_LABELS[standard_gate as usize].to_string()
```

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+                let custom_label = instruction.label.clone();
+                if custom_label.is_some() && *custom_label.unwrap() != label {
+                    label = *instruction.label.clone().unwrap()
+                }
```
Could this just be:
```suggestion
                if let Some(custom_label) = instruction_label {
                    if custom_label != label {
                        label = custom_label.clone();
                    }
                }
```

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+                        .params_view()
+                        .iter()
+                        .map(|param| match param {
+                            Param::Float(f) => format!("{}", f),
```
I think you should be able to just do:
```suggestion
                            Param::Float(f) => f.to_string(),
```

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+            OperationRef::Unitary(_) => *instruction
+                .label
+                .clone()
+                .unwrap_or(Box::new(" Unitary ".to_string())),
```
Why does this need to be in a box?

Shobhit21287 replied:
The label stored inside `PackedInstruction` is of the form `Option<Box<String>>` because of which when doing `unwrap_or` the compiler forces the 'or' option to be `Box<String>`. Is there a way to get around this?

mtreinish replied:
You shouldn't need to allocate a box for this, since you're working with an owned string. You just need to copy the contents of the boxed label instead of cloning the outer box.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/15357/commits/c28a146ebb6e6eb1b00df180c8645bfd23e5f1f2

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+        for (i, element) in layer.0.iter().enumerate() {
+            let wire = Self::draw_element(element, vis_mat, cregbundle, i);
+            wires.push(wire);
+        }
```
Can't you just do this?
```suggestion
        let wires: Vec<_> = layer.0.iter().enumerate().map(|(i, element)| Self::draw_element(element, vis_mat, cregbundle, i)).collect();
```

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+            if w > layer_width {
+                layer_width = w;
+            }
+        }
```
Is this equivalent to?:
```suggestion
        let layer_width = wires.iter().map(|w| w.width()).max().unwrap_or(0);
```

Shobhit21287 replied:
done in [58ca1f7](https://github.com/Qiskit/qiskit/pull/15357/commits/58ca1f724064a0fb9595de84aaa596ee6250eee0)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+        ind: usize,
+    ) -> TextWireElement {
+        let circuit = vis_mat.circuit;
+        let (top, mid, bot);
```
A small style thing, I probably would have just made things return the `TextWireElement` in all the branches directly. They all individually need to set all 3 in all cases but instead of just doing         `TextWireElement { top, mid, bot }` at the very end you could remove the semicolon on the match and have it return the `TextWireElement` from each possible match return. I don't think it matters too much, it just is a bit uncommon to see the unassigned declaration like this in Rust, especially for something like this. But since it compiles I think the compiler is fully checking that every branch assigns all 3 so it's probably fine.

eliarbel replied:
In earlier versions of this function, it was actually written the way you suggest here, but then we switched to the current approach to reduce the level of nesting in this code, which is still not nice, but at least better than before.

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+                current_fold = layer_width;
+            }
+        }
+        ranges.push((start, num_layers));
```
I think you could do this with a iterator collect over `layer_widths` using `filter_map()` and a `chain()` for the last push outside the loop.

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+        for (start, end) in ranges {
+            let mut wire_strings: Vec<String> = Vec::new();
+
+            for wire in &self.wires {
```
It's a good habit to get into to use iterator collections for thins like this with fixed sizes. It arguably makes it a bit easier to read but also enables to compiler to optimize things a bit more typically around preallocating the vecs. You can get most if not all the performance from by just using `with_capacity()` instead of `new()` if the for loop is easier to work with for some reason.

mtreinish commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,1247 @@
+        let mut output = String::new();
+
+        for (start, end) in ranges {
+            let mut wire_strings: Vec<String> = Vec::new();
```
This changes things for the comment below but in this specific case it would be more efficient to reuse the allocation of the vec for each iteration. Something like:
```suggestion
        let mut wire_strings: Vec<String> = Vec::new();
        for (start, end) in ranges {
            wire_strings.clear();
```
will a lot of overhead from allocation and deallocations.

Shobhit21287 replied:
done in[5cca97d](https://github.com/Qiskit/qiskit/pull/15357/commits/5cca97dca74a6fb7ee598cf0399cdfe193e01d61)

mtreinish reviewed (commented):
Overall this looks really good, I like how you organized the drawer here the idea of defining a grid of elements that you represent how everything gets drawn which you then just translate into text as the last step makes a lot of sense and makes it really easy to follow.

I left a few comments inline most are just mechanical or suggestions on rust syntax but a few there are some potential longer term issues (like the way the index mapping is done between the dag and circuit). A general comment in the code mechanics I know kept saying in the inline comments that performance for the visualizer doesn't matter and in practice it doesn't. But as I've trained myself to look at it for other places in Qiskit the main feeling I had while reviewing is that there are a lot of allocations going on throughout the code (especially in the text manipulation sections) and my feeling was that a fair number of them could be removed. I didn't call out all the places I saw it, but it might be worth going over how you use `String`, `Vec`,  `format!()`, and other calls/types that will heap allocate to see if you really need to allocate there.

The only large scale question I have was how are you intending to scope this new function? Right now there is a private python function being added to call this new drawer but the PR summary say the intent is not to replace the python drawer. If that's the case I'm assuming this is just for the C API, then I'm not sure we need to expose this to python.

The other thing that would be really good to see here is some testing of this new code. As a general rule we want any code being added to Qiskit to have coverage in CI to verify it's working correctly not just at the time of merging but also for any future changes to catch any regressions. However right now there is nothing exercising these code paths in any of our test suites. You should be able to write rust tests that validate circuits are  generating expected strings. In general we can't really merge a large function like this without having test coverage.

Shobhit21287 committed:
used iterators in multiple places to make the code efficient, inplace allocation to prevent intermediate strings

Shobhit21287 committed:
replaced a for loop with iterator, linting and final checks

Shobhit21287 committed:
added tests for merge_wire, cregbundle and folding

mtreinish committed:
Merge remote-tracking branch 'origin/main' into text_circuit_drawer

mtreinish committed:
Cleanup tests and strip trailing whitespace

This commit updates the formatting in the unit tests to make it easier
to maintain. It uses multi-line strings in the source to show the output
of the visualizations as they would be printed to make it easier to
debug. Additionally, one aspect of the output that made working with it
harder was all the lines in the strings were fixed width with
whitespace. This is an artifact of the matrix approach that was filling
in each space in the visualization even if it didn't have anything to
draw in that coordinate. This was making the visualization tests harder
to debug because there was a lot of whitespace that had to be exactly
right. To address this and to make the output strings easier to work
with this commit strips the unnecessary whitespace from the output.

mtreinish committed:
Update use for crossterm to indicate source of size() call

This commit updates the crossterm::terminal::size() use statement to
avoid a bare `size()` function call. It wasn't super clear from reading
that this is referring to the terminal size as "size" is a fairly
generic function name. This switches to importing the `terminal` module
making the sole call to the function `terminal::size()` which provides a
bit more context around the call making the code clearer.

mtreinish committed:
Remove the python entry point function

This commit removes the pyfunction that enables calling the rust drawer
from python. This is not needed right now becase the scope of this
function is to be a drawer for C. In the future if it reaches feature
pairity with the Python we can expose it to python and replace the
python drawer's code with this. But, for this initial implementation we
don't need to expose it to Python.

mtreinish committed:
Remove TODO comment

This commit removes a TODO comment about renaming a struct. This wasn't
needed because the name is fine, especially as it's in use in the
current code for the drawer. We can rename it later, but it's not really
a TODO item at this point, especially as we prepare to merge this PR.

mtreinish committed:
Avoid box creation in unitary label handling

mtreinish committed:
Use label for fallback

This commit updates the fallback operation label in the output
visualization to first use the label field if set and then the operation
name as a fallback if a label is not set.

mtreinish committed:
Add missing whitespace back to fallback name string

The previous commit erroneously removed the whitespace around the string
label set when using the name. This restores that extra whitespace
padding.

mtreinish committed:
Add global phase display and expand test coverage

This commit expands the test coverage to cover more of the operating
space of this function. We want to be able to cover the surface area of
the types of circuits coming from C as that is the intent of this
function. This commit expands the test coverage to include more of this
surface area, although it's probably still incomplete.

The drawer was missing the global phase from the visualization. This
commit adds it so that when visualizing a circuit users can see the
global phase. This was found during the expansion of the test coverage
for setting global phase.

mtreinish committed:
Skip miri on tests with dynamic terminal width

This commit adds miri skips on the tests that are using the
circuit_drawer with dynamic terminal width determination. This relies on
calls miri can't track (using the fs) and will cause miri to fail.
Skipping is the only option in these cases as we want coverage of these
code paths.

mtreinish committed:
Update copyright header

In #15629 the copyright header url was updated to use https instead of
http. This was not reflected in this PR though, this commit makes the
necessary update.

mtreinish committed:
Simplify range loop

mtreinish committed:
Update barrier usage in tests

mtreinish committed:
Fix parameterized output

This commit fixes the display of ParameterExpression parameters for
global phase and gate/instruction parameters. Previously it was using
debug formatting for parameter expressions which was resulting in
debugging output instead of a normal string representation of the
expression. This switches to use the ParameterExpression::to_string()
method which is more useful.

mtreinish commented:
I'm working on adding more test coverage, and I've found a formatting issue that I'm still trying to debug. Printing `RXX(phi) - RXX(phi) - RZX(2)` (and the second rxx has a label) results in:

```
     ┌───────────┐┌──────────────┐┌──────────┐
q_0: ┤0  Rxx(ϕ) ├┤0  my_rxx(ϕ) ├┤0  Rzx(2) ├
     │           ││              ││          │
     │          │ │             │ │          │
q_1: ┤1         ├─┤1            ├─┤1         ├
     └──────────┘ └─────────────┘ └──────────┘
```
The alignment for the top and bottom section on the box with the label is too long.

mtreinish committed:
Fix unicode width handling in drawer

This commit fixes the handling of unicode character widths in the
drawer. Previously the width of strings was determined by
`string.chars().count()` which was only valid for ascii and a subset of
UTF-8 as it's counting the number of bytes not even the unique graphemes
as in UTF-8 graphemes are variable length. To address this we need to
use external crates, first unicode-width is used to estimate the number
of columns of text are used to display the graphemes in a string. This
is needed because some unicode is wider than a single column, especially
some of the emoji and icons. When calculating how wide to make boxes in
the drawings this is important because we need to know how long to make
the boxes to match the rendered width of the labels. The second crate is
unicode-segmentation is used to provide an iterator over graphemes
instead of raw bytes.

mtreinish commented:
I've fixed the unicode width handling in: https://github.com/Qiskit/qiskit/pull/15357/commits/4940b85cb852d9703a6add34f127650b4650f569 and added some unit tests for it.

mtreinish committed:
Build visualization layers on circuit data

This commit adjusts the circuit drawer to avoid the intermediate
DAGCircuit usage. Previously a DAGCircuit was constructed to compute
layers of gates for drawing. The downside with this besides the general
overhead is that we need to keep the instructions in circuit data mapped
to the dagcircuit which complicates the code. However, this wasn't
strictly needed as we can generate layers directly from a circuit data.
This commit makes this change and adds an algorithm for computing layers
by iterating over circuit data. This algorithm does not produce layers
in the same way as DAGCircuit would, it is sensitive to insertion order
and maybe produce different layers depending on the order which gates
are added to the circuit. However as this drawer is just for C/Rust to
start and this is an initial implementation we can change and improve
this over time. The initial algorithm added in this commit is fairly
simple.

mtreinish committed:
Merge branch 'main' into text_circuit_drawer

eliarbel commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,2003 @@
+        }
+    };
+    // Strip trailing whitespace from lines.
+    // On the last line ensure we only a single newline ends the returned
```
```suggestion
    // On the last line ensure only a single newline ends the returned
```

eliarbel replied:
Done in 6f4f08d52e1ec016288f000e8d325f9077f93b35

eliarbel commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,2003 @@
+    };
+    // Strip trailing whitespace from lines.
+    // On the last line ensure we only a single newline ends the returned
+    // string (in case we ended up with a double newline after the stripping)
```
The comment better fits before L75 (`output.extend(`)

eliarbel replied:
Done in 6f4f08d52e1ec016288f000e8d325f9077f93b35

eliarbel commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,2003 @@
+            std::mem::swap(&mut layer, &mut new_layer);
+            layers.push(new_layer);
+            bits.iter_mut().enumerate().for_each(|(idx, x)| {
+                *x = idx >= node_min || idx <= node_max;
```
I guess you meant it to be a conjunction:
```suggestion
                *x = node_min <= idx && idx <= node_max;
```

eliarbel replied:
layering algorithm has been updated in d04592180dddc9df4e4be0fcec3a109375ce81c3, so this is not relevant anymore

eliarbel commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,2003 @@
+        );
+        if (node_min..=node_max).any(|idx| bits[idx]) {
+            let mut new_layer = vec![inst];
+            std::mem::swap(&mut layer, &mut new_layer);
```
I wonder why we need the `swap` here, instead of first pushing the current layer to `layers` and then binding a new vector like `layer = vec![inst]`?

eliarbel replied:
Not relevant anymore after d04592180dddc9df4e4be0fcec3a109375ce81c3

eliarbel commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,2003 @@
+        inst: &'a PackedInstruction,
+        circuit: &CircuitData,
+    ) {
+        if gate == StandardGate::GlobalPhase {
+            return;
+        }
```
Just a comment about style and it's minor anyway so feel free to ignore: Since the `StandardGate::GlobalPhase` arm below is a no-op anyway, I read this is an optimization to skip the 2 lines that come immediately after. Personally, I'd choose code simplicity here over this as optimization which practically I'd expect to not help much (though also not hurt really with branch prediction).

mtreinish replied:
I actually didn't add it as an optimization. The reason was the `get_instruction_range()` method doesn't work with a zero length qargs list. So I have to handle this case with a condition somewhere to avoid trying to find the range for 0 length args. It seemed simpler to do this, then do the check in `get_instruction_range()`. I can add a comment to this effect to explain the reasoning.

eliarbel replied:
Oh, I see. That was an oversight when I implemented `get_instruction_range()`. Ok, then that if statement is actually for correctness. I can update `get_instruction_range()` to support `GlobalPhaseGate` by returning the `(usize, usize)` as `Some`, but let's do that in a follow-up PR.

eliarbel commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,2003 @@
+    pub fn merge_lines(top: &str, bot: &str) -> String {
+        let mut ret = String::new();
+
+        for (topc, botc) in top.graphemes(true).zip(bot.graphemes(true)) {
```
TIL graphemes 🏫

mtreinish replied:
Yeah it was a new word for me too when I started looking at this library. I was more surprised the standard library didn't have a way to do this, but the guidance I found was to use this library.

eliarbel commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,2003 @@
+    }
+}
+
+#[cfg(test)]
```
The `py_drawer` function which was removed in aa9db91ad5d12171739ef26b6414ea8294c03387 was really handy with testing and comparing the Rust drawer against the Python one during development. Maybe we should keep it a bit longer? We can rename the Python name to maybe `_draw` to highlight it's not meant to be public.

mtreinish replied:
I'm not sure I see the value anymore. I can understand it was useful during bring-up but with a rust test suite now and #15361 adding it to the C API I don't think it's worth exposing as a hidden internal function anymore. I didn't need Python during the work I did over the last week or two on this PR. Part of my hesitation to add it back here is that this is just one more place that pulls in pyo3 in the core circuit crate and we're trying (although not succeeding yet) to get better about isolating python in the rust code.

It'd be different if we were considering using the rust drawer for Python, but I didn't think that was in scope in the near term. That being said we're probably in a place now with the rust data model where we could replace the python drawer with the rust implementation. The only thing not in rust currently are python defined custom instructions, but all we need for those are the name and qargs/cargs. So it's definitely in the cards medium term.

eliarbel replied:
Correct, using the Rust drawer for Python is not in scope for this PR, since it was built with the C API in mind so things like for control flow and `PauliProductMeasurement` are not supported yet. 

Ok, adding the Python function for debug during development is trivial so if you prefer to save the PyO3 dependency here that's fine.

eliarbel commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,2003 @@
+    }
+
+    #[test]
+    fn test_fold() {
```
Some of the tests are not excluded for `miri`. Should we have the module itself attributed with `#[cfg(not(miri))]`?

mtreinish replied:
The `not(miri)` condition is only needed for the tests that are not setting a fixed terminal width. In those cases the code calls out to figure out the terminal width and miri doesn't work with that. With a fixed width set we can run the drawer under miri just fine though. I left a couple of tests in place that didn't set a fixed width because I wanted to exercise that code path of width discovery (I chose relatively shallow circuits on purpose to avoid the risk of a small terminal folding causing a failure) but otherwise tried to set a fixed width to be able to run things under miri.

eliarbel commented on crates/circuit/src/circuit_drawer.rs:
```diff
@@ -0,0 +1,2003 @@
+«       │   │           │
+«     ┌─┴──┐│           │
+«q_3: ┤ Sx ├┤ 3         ├
+«     └────┘└───────────┘
```
Nitpick, fine if ignored: it would be nice if we can make the expected result drawing denser by squeezing the gates modulo qubit indices. E.g. in the lines of the C test in #15361: 
```c
uint32_t qubits[10] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
for (uint8_t gate = 0; gate <= QkGate_RC3X; ++gate) {
        qk_circuit_gate(circuit, gate, &qubits[gate % 8], params);
    }
```

eliarbel replied:
Handled in d04592180dddc9df4e4be0fcec3a109375ce81c3

eliarbel reviewed (commented):
Thanks Matthew for contributing to this PR and helping with pushing it to the finish line. I've reviewed your contributions only, i.e. from c498cfeaf6e48f4daac418e6f3caae0ecdea78f6 onward. 

I had one important comment in `build_layers` and additional minor ones in several other places. 

My additional general comment is regarding the new `CircuitData`-based layering algorithm (in `build_layers`). It's seems to be working OK however it's too eager with adding new layers. For example, this Python code:
```python
for i in range(3):
    qc.h(0)
for i in range(3):
    qc.h(1)
```

results with this drawing: 
```text
           ┌───┐┌───┐┌───┐
   MyQ1_0: ┤ H ├┤ H ├┤ H ├──────────
           └───┘└───┘├───┤┌───┐┌───┐
   MyQ1_1: ──────────┤ H ├┤ H ├┤ H ├
                     └───┘└───┘└───┘
```
while the Python drawer gives this: 
```text
          ┌───┐┌───┐┌───┐
  MyQ1_0: ┤ H ├┤ H ├┤ H ├
          ├───┤├───┤├───┤
  MyQ1_1: ┤ H ├┤ H ├┤ H ├
          └───┘└───┘└───┘
```

I think it would be good if we can modify `build_layers` so that it stacks gates/instructions more densely. I think that if instead of using a boolean vector (`bits`) to decide when to creata a new layer, we track the frontier of layers across the wires based on the qargs/cargs ranges of the instructions as we iterate through them, we would be able to map each instruction to the leftmost layer without collision. If you're busy I'll be happy to push a commit.

eliarbel committed:
Merge remote-tracking branch 'upstream/main' into text_circuit_drawer

eliarbel committed:
Make build_layers less eager with adding new layers

eliarbel committed:
Fix rendering bug of controlled gates

Fixes a bug where controlled gates (e.g. a CPhase gate) had
their top or bottom connectors misaligned by one position with
the corresponding controlled lines, in case their label width
is even.

eliarbel commented:
Following an offline discussion, I pushed a few commits in hope this will get merged in time for the 2.4 release. 
The changes per commit are: 
- https://github.com/Qiskit/qiskit/pull/15357/commits/6177067e307459737aac3896e7e4195f1fe6511b - resolving a conflict
- https://github.com/Qiskit/qiskit/pull/15357/commits/d04592180dddc9df4e4be0fcec3a109375ce81c3 - updating -`build_layers` logic per this comment: https://github.com/Qiskit/qiskit/pull/15357#pullrequestreview-3924609287. This also addresses this comment:  https://github.com/Qiskit/qiskit/pull/15357#discussion_r2913845630
- https://github.com/Qiskit/qiskit/pull/15357/commits/9a00b5a2ebff0da9b7bf3ff5a8e5bd05823855f0 - fixing a small rendering bug which resulted in output like this: 
```text
├──────■──────┤ 
       │ 
┌─────┴─────┐   
┤ P(3.141)  ├─
└───────────┘ 
```
@alexanderivrii  FYI

eliarbel committed:
Update comment for connector alignment

eliarbel committed:
Address remaining minor comments

mtreinish reviewed (approved):
Thanks for finishing this up Eli, I think between both of our reviews and the recent fixes this is definitely good to merge now. Thanks @Shobhit21287 for all the work on this.
